aten/src/ATen/native/cuda/CUDALoops.cuh

```diff
@@ -129,17 +129,22 @@ static inline void launch_unrolled_kernel(int64_t N, const func_t& f, array_t da
 }
 
 template<int nt, int vt, typename func_t>
-C10_LAUNCH_BOUNDS_2(nt, 4)
+C10_LAUNCH_BOUNDS_2(nt, 2)
 __global__ void elementwise_kernel(int N, func_t f) {
-  int tid = threadIdx.x;
-  int nv = nt * vt;
-  int idx = nv * blockIdx.x + tid;
+  // vt is elements unroll per thread
+  int start_idx = vt * (threadIdx.x + blockIdx.x * blockDim.x);
+  int grid_stride = vt * (gridDim.x * blockDim.x);
+
   #pragma unroll
-  for (int i = 0; i < vt; i++) {
-    if (idx < N) {
-      f(idx);
-      idx += nt;
-    }
+  for (int i = start_idx; i < N; i += grid_stride) {
+    # pragma unroll
+    for (int k = 0; k < vt; k++)
+      f(i+k);
+  }
+  if (N % vt != 0 && threadIdx.x == 0 && blockIdx.x == 0) {
+    f(N-1);
+    f(N-2);
+    f(N-3);
```

Comment:
Is this faster than doing something like:

``` cuda
if (threadIdx.x == 0 && blockIdx.x == 0) {
# pragma unroll
  for (i = 1; i <= N % vt; i++) {
    f(N - i);
  }
}
```

The current code assumes `vt == 4` (which is currently true, but may potentially change in the future).

Reply:
Fixed without the unroll on the loop.